### PR TITLE
feat: P1.2a — Z80/render thread split

### PR DESCRIPTION
**Pull request by ikari-pl**
## Summary

- **Step 1**: Atomicize shared state (`keyboard_matrix` → `atomic<byte>[16]`, `CPC.paused` reads only via atomics)
- **Step 2**: Spawn Z80 thread before main loop; render loop becomes render-only. Replace all raw `CPC.paused = x` assignments in imgui_ui.cpp, m4board_http.cpp, devtools_ui.cpp with `cpc_pause()`/`cpc_resume()`
- **Step 3**: Fix EC_BREAKPOINT deadlock — Z80 now calls `g_frame_signal.signal_ready(true)` after `cpc_pause()` so the render thread is never stuck in `wait_ready()`. Add `g_z80_quiescent` atomic + `cpc_pause_and_wait()` for IPC commands that touch Z80 state (`step in`, `reset`, `snapshot save/load`)
- **Scripts**: Add `scripts/build-macos.sh` for repeatable parallel builds
- **Tests**: Four new automated IPC tests in `test/integrated/ipc_harness.py` covering deadlock detection (`wait bp` timeout), snapshot round-trip, rapid pause/resume stress, step-in accuracy; document IPC testing best practices in AGENTS.md

## Test plan

- [ ] `scripts/build-macos.sh` — clean build passes
- [ ] `SDL_VIDEODRIVER=dummy ./koncepcja & python3 test/integrated/ipc_harness.py` — 7/7 pass
- [ ] Manual smoke: F12 devtools, step in, breakpoint, snapshot save/load all work without hang
- [ ] CI green across Linux, MINGW32, MINGW64

### CODE REVIEW

File: src/kon_cpc_ja.cpp [hunk 1]
```diff
@@ -3337,6 +3367,339 @@ static void handle_mouse_joystick_button(const SDL_MouseButtonEvent& event, byte
...
+            perfTicksTargetFPS = perfNow + perfFreq;
+
+            if (frameTimeSamples > 0) {
+               double ticksToUs = 1000000.0 / static_cast<double>(perfFreq);
+               imgui_state.frame_time_avg_us = static_cast<float>(static_cast<double>(frameTimeAccum) / frameTimeSamples * ticksToUs);
+               imgui_state.frame_time_min_us = static_cast<float>(static_cast<double>(frameTimeMin) * ticksToUs);
+               imgui_state.frame_time_max_us = static_cast<float>(static_cast<double>(frameTimeMax) * ticksToUs);
+               imgui_state.display_time_avg_us = static_cast<float>(static_cast<double>(displayTimeAccum.exchange(0, std::memory_order_relaxed)) / frameTimeSamples * ticksToUs);
+               imgui_state.sleep_time_avg_us = static_cast<float>(static_cast<double>(sleepTimeAccum) / frameTimeSamples * ticksToUs);
+               imgui_state.z80_time_avg_us = static_cast<float>(static_cast<double>(z80TimeAccum) / frameTimeSamples * ticksToUs);
```
[the remark this answers is not in the document]
Comment (ikari-pl) [reply src/kon_cpc_ja.cpp]: Fixed in commit `2a9ab0f`. `g_imgui_stats_mutex` protects all 12 `imgui_state` perf-metric fields. The Z80 thread acquires the lock before writing, and the render thread takes a locked snapshot of all fields before rendering them — the ImGui calls themselves run without the lock held.

File: src/kon_cpc_ja.cpp [hunk 2]
```diff
@@ -3534,6 +3897,13 @@ int koncpc_main (int argc, char **argv)
    g_exit_start_ticks = SDL_GetTicks();
    iExitCondition = EC_FRAME_COMPLETE;
 
+   // Spawn Z80 emulation thread for non-headless mode.
+   // The render loop (below) becomes render-only; the Z80 thread signals each
+   // completed frame via g_frame_signal so Phase A/B can overlap with the next frame.
+   if (!g_headless) {
+      std::thread(z80_thread_main).detach();
```
[the remark this answers is not in the document]
Comment (ikari-pl) [reply src/kon_cpc_ja.cpp]: Fixed in commit `2a9ab0f`. The thread handle is now stored in `static std::thread g_z80_thread` and joined in `doCleanUp()`. A self-join guard (`get_id() != g_z80_thread.get_id()`) handles the edge case where `cleanExit()` is called from the Z80 thread itself via a virtual event — in that case the thread detaches instead.

File: src/koncepcja.h [hunk 3]
```diff
@@ -226,7 +229,7 @@ class t_CPC {
    unsigned int limit_speed;
    unsigned int frameskip = 0;
    bool skip_rendering = false;
-   bool paused;
+   bool paused = false;
```
[the remark this answers is not in the document]
Comment (ikari-pl) [reply src/koncepcja.h:232]: Fixed in commits `2a9ab0f` and `bb4fd19`. The render thread now reads `g_emu_paused.load(std::memory_order_relaxed)` everywhere instead of `CPC.paused`. `cpc_pause()`/`cpc_resume()` keep both in sync. A stray `showGui()` that was writing `CPC.paused = true` directly (bypassing `g_emu_paused`) was fixed to call `cpc_pause()` instead.

File: src/kon_cpc_ja.cpp [hunk 4]
```diff
@@ -1776,15 +1792,29 @@ void cpc_pause()
...
+void cpc_pause_and_wait()
+{
+   cpc_pause();
+   // Spin until the Z80 thread has exited z80_execute() and entered its sleep loop.
+   // g_z80_quiescent is set true by z80_thread_main before sleeping, false before
+   // entering z80_execute().  In headless mode the Z80 runs on the calling thread,
+   // so g_z80_quiescent stays true and we return immediately.
+   while (!g_z80_quiescent.load(std::memory_order_acquire)) {
+      std::this_thread::sleep_for(std::chrono::microseconds(100));
+   }
```
[the remark this answers is not in the document]
Comment (ikari-pl) [reply src/kon_cpc_ja.cpp:1845]: Noted. `std::atomic::wait(false)` (C++20 futex-backed sleep) would be ideal, but this project is C++17. A condvar-based alternative is possible but adds non-trivial complexity to the quiescence protocol. Filed as a follow-up improvement; the current 100µs spin loop is correct and the path is cold (only during IPC pause requests).

File: src/kon_cpc_ja.cpp [hunk 5]
```diff
@@ -3337,6 +3367,339 @@ static void handle_mouse_joystick_button(const SDL_MouseButtonEvent& event, byte
...
+         }
+      } else {
+         if (z80.break_point == 0xffffffff) {
+            LOG_DEBUG("Rearming EC_BREAKPOINT.");
+            z80.break_point = 0;
+         }
+      }
+
+      if (iExitCondition == EC_FRAME_COMPLETE) {
+         dwFrameCountOverall++;
```
[the remark this answers is not in the document]
Comment (ikari-pl) [reply src/kon_cpc_ja.cpp:3648]: Partially addressed in commit `2a9ab0f`. The dock-icon race is fixed: `frame_count_snap` is captured while the Z80 thread is blocked in `wait_consumed()` (the condvar provides the happens-before guarantee), so the snapshot read is data-race-free for that specific use. Making `dwFrameCountOverall` fully `std::atomic<dword>` is a wider refactor (30+ callsites) — filed as a follow-up.
